stratum/hal/lib/common/phaldb_service.cc

```diff
@@ -0,0 +1,454 @@
+    results = {};
+
+    // Do set for all attribute pairs
+    auto adapter =
+        absl::make_unique<::stratum::hal::phal::Adapter>(phal_interface_);
+    status = adapter->Set(attrs);
+  }
+
+  return ToGrpcStatus(status, results);
+}
+
+::grpc::Status PhalDBService::Subscribe(::grpc::ServerContext* context,
+    const ::stratum::hal::phal::SubscribeRequest* req,
+    ::grpc::ServerWriter<::stratum::hal::phal::SubscribeResponse>* stream) {
+
+  RETURN_IF_NOT_AUTHORIZED(auth_policy_checker_, PhalDBService,
+                           Subscribe, context);
+
+  ::util::Status status;
+  ::util::StatusOr<::stratum::hal::phal::Path> result;
+
+  // Convert to PhalDB Path
+  switch (req->query_case()) {
+    case ::stratum::hal::phal::SubscribeRequest::kStr: {
+        result = ParseQuery(req->str());
+        break;
+    }
+    case ::stratum::hal::phal::SubscribeRequest::kPath: {
+        result = ToPhalDBPath(req->path());
+        break;
+    }
+    default:
+        return ::grpc::Status(
+            ::grpc::StatusCode::INVALID_ARGUMENT,
+            "Invalid query in Subscribe request.");
+  }
+
+  // Save status
+  status = result.status();
+
+  // If conversion worked
+  if (result.ok()) {
+    auto path = result.ConsumeValueOrDie();
+
+    // Create writer and reader channels
+    std::shared_ptr<Channel<::stratum::hal::phal::PhalDB>> channel =
+        Channel<::stratum::hal::phal::PhalDB>::Create(128);
+
+    {
+        // Lock subscriber channels
+        absl::MutexLock l(&subscriber_thread_lock_);
+
+        // Save channel to subscriber channel map
+        subscriber_channels_[pthread_self()] = channel;
+    }
+
+    auto writer =
+            ChannelWriter<::stratum::hal::phal::PhalDB>::Create(channel);
+
+    auto reader =
+        ChannelReader<::stratum::hal::phal::PhalDB>::Create(channel);
+
+    // Issue the subscribe
+    auto adapter =
+        absl::make_unique<::stratum::hal::phal::Adapter>(phal_interface_);
+    status = adapter->Subscribe({path}, std::move(writer),
+                                 absl::Seconds(req->polling_interval()));
+
+    // If Subscribe ok
+    if (status.ok()) {
+        // Loop around processing messages from the PhalDB writer
+        // Note: if the client dies we'll only close the channel
+        //       and thus cancel the PhalDB subscription once we
+        //       get something from the PhalDB subscription (i.e.
+        //       if the poll timer expires and something has changed).
+        //       We could potentially put something in here to check
+        //       the stream and channel for changes but for now this
+        //       will do.
+        do {
+            ::stratum::hal::phal::PhalDB phaldb_resp;
+            int code = reader->Read(&phaldb_resp,
+                                    absl::InfiniteDuration()).error_code();
+
+            // Exit if the channel is closed
+            if (code == ERR_CANCELLED) {
+                status = MAKE_ERROR(ERR_INTERNAL)
```

Comment:
Don't overwrite the previous status. `APPEND_STATUS_IF_ERROR` collects all of them.

Example: https://github.com/stratum/stratum/blob/a3e56b36d8a312e19ab929d0b39814382f38dd26/stratum/hal/lib/bcm/bcm_packetio_manager.cc#L259